### DISCUSSION TITLE
How do I train an autoencoder on spark Pytorch

### DISCUSSION THREAD

**Discussion by xiaodaigh** [horovod/horovod#2760] [Q&A]
**Environment:**
1. PyTorch on Spark
2. Framework version: Spark v3 pytorch1.6
3. Horovod version: horovod 0.20.3
4. MPI version: Not sure
5. CUDA version: Not sure
6. NCCL version: Not sure
7. Python version: 3.7
8. OS and version: Ubuntu 18.04
9. GCC version: not sure

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes
2. If your question is about hang, did you read: NA
3. If your question is about docker, did you read NA
4. Did you check if you question is answered in the [troubleshooting guide] Yes

**Your question:**
I know that I can use `horovod.TorchEstimator` to  fit a model and then use `torch_estimator.fit(df)` to create the outcome flag.

Like this
```
torch_estimator = hvd.TorchEstimator(#num_proc=2, #args.num_proc,
                                        store=store,
                                        model=model,
                                        optimizer=optimizer,
                                        loss=lambda model_output, target: loss(model_output, target.long()),
                                        input_shapes=[[-1, 1]]*11,
                                        feature_cols=feature_cols,
                                        label_cols=target_cols,
                                        #batch_size=32, #args.batch_size,
                                        epochs=2, #args.epochs,
                                        #verbose=1
                                        )

torch_model = torch_estimator.fit(spark_df).setOutputCols(['label_prob'])
```
But the an autoencoder has a special structure in that I have an `encoder` and `decoder`. When fitting the model, I want to fit both the encoder and decoder networks, however, when `transform` I only want to apply the `encoder`.

Like this

```
class Autoencoder(nn.Module):
    def __init__(self):
        super().__init__()
        
        self.encoder = nn.Sequential(
            nn.Linear(n1, n2),
            nn.ReLU(True),
            nn.Linear(n2, n3)

        self.decoder = nn.Sequential(
            nn.Linear(n3, n2),
            nn.ReLU(True),
            nn.Linear(n2, n1),
            nn.Tanh()) 


    def forward(self, data):
        """ This is the forward computation of taking as input the data
            and outputing the neural network approximated input """

         x = torch.cat(data, 1)

        return self.decoder(self.encoder(x))
```

However, I only see the `transform` method like `torch_model.transform(spark_test_df)` method. Is it possible to use `transform` but only using the `encoder`?

**Comment by tgaddair** [comment]
Hey @xiaodaigh, good question.  You will be able to do this once #2316 lands, which adds support for `LightningModule` inputs to the `TorchEstimator`.
